EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+        )
+        
+        # Process transaction and update block-level sets
+        gas_used, accessed_addresses, accessed_storage_keys, logs, error = process_transaction(env, tx)
+        block_level_accessed_addresses += accessed_addresses
+        block_level_accessed_storage_keys += accessed_storage_keys
+```
+
+This code shows how block-level slot warming is implemented at the execution layer. The `block_level_accessed_addresses` and `block_level_accessed_storage_keys` sets are maintained throughout the block's execution and passed to each transaction's environment.
+
+#### Transaction Processing
+
+```python
+def process_transaction(env: vm.Environment, tx: Transaction) -> Tuple[Uint, Tuple[Log, ...], Optional[Exception]]:
+    preaccessed_addresses = set()
+    preaccessed_storage_keys = set()
+    
+    # Add block-level pre-accessed slots
+    preaccessed_addresses.add(env.block_level_accessed_addresses)
+    preaccessed_storage_keys.add(env.block_level_accessed_storage_keys)
+    
+    # Handle access lists from transaction
+    ...
+```
+
+This adds the block-level-accessed addresses and storage keys to the preaccessed addresses and storage keys.
+As a result, from the perspective of a transaction, block-level accessed addresses and storage keys are treated the same as precompiles or the coinbase address.
+
+```python
+def process_message_call(message: Message, env: Environment) -> MessageCallOutput:
+    return MessageCallOutput(
+        # ... other fields ...
+        accessed_addresses=evm.accessed_addresses,
+        accessed_storage_keys=evm.accessed_storage_keys
+    )
+```
+
+The message call processing tracks accessed addresses and storage keys during execution, which are then propagated back up to the transaction level and ultimately to the block level.
+
+## Rationale
+
+The proposal builds on several key observations:
+
+1. **Caching Efficiency**: Modern Ethereum clients already implement sophisticated caching mechanisms at the block level. Extending address and storage key warming to match this caching behavior better aligns gas costs with actual computational costs.
+
+2. **Backward Compatibility**: The worst-case scenario for any transaction remains unchanged - it will never pay more than the current cold access cost for its first access to a slot.
+
+3. **First Access Warms For All**: Under the proposed mechanism, the first transaction in a block that accesses and warms multiple addresses or storage slots bears the entire cost of warming, even if subsequent transactions benefit from those already-warmed slots without incurring additional costs. This approach is deemed acceptable because early execution within a block—where warming typically occurs—is generally occupied for transactions placed by professional builders aiming for top-of-block execution. Since the resulting cost discrepancy is relatively minor, a more complex cost-sharing model is avoided in favor of maintaining simplicity.
```

Comment:
This is an interesting comment, since competition will thus now also be much more dependent upon what storage was added and what not. Therefore to pack more logic / more transactions this thus means that transactions which interact with already warm addresses/slots are more preferred than those who are not. (This would be an interesting research also, because this could theoretically mean that (in an extreme example) for block A we will only see txs with contract Y, and then for block B only those for block Z, but this obviously also depends upon the goal/optimization target of the builder)